src/dbnode/encoding/proto/custom_unmarshal.go

```diff
@@ -0,0 +1,379 @@
+	switch wireType {
+	case proto.WireFixed32:
+		num, err := u.decodeBuf.decodeFixed32()
+		if err != nil {
+			return zeroValue, err
+		}
+		return unmarshalSimpleField(fd, num)
+	case proto.WireFixed64:
+		num, err := u.decodeBuf.decodeFixed64()
+		if err != nil {
+			return zeroValue, err
+		}
+		return unmarshalSimpleField(fd, num)
+	case proto.WireVarint:
+		num, err := u.decodeBuf.decodeVarint()
+		if err != nil {
+			return zeroValue, err
+		}
+		return unmarshalSimpleField(fd, num)
+
+	case proto.WireBytes:
+		if fd.GetType() != dpb.FieldDescriptorProto_TYPE_BYTES &&
+			fd.GetType() != dpb.FieldDescriptorProto_TYPE_STRING {
+			// This should never happen since it means the skipping logic is not working
+			// correctly or the message is malformed since proto.WireBytes should only be
+			// used for fields of type bytes, string, group, or message. Groups/messages
+			// should be handled by the skipping logic (for now).
+			return zeroValue, fmt.Errorf(
+				"tried to unmarshal field with wire type: bytes and proto field type: %s",
+				fd.GetType().String())
+		}
+
+		// Don't bother copying the bytes now because the encoder has exclusive ownership
+		// of them until the call to Encode() completes and they will get "copied" anyways
+		// once they're written into the OStream.
+		raw, err := u.decodeBuf.decodeRawBytes(false)
+		if err != nil {
+			return zeroValue, err
+		}
+
+		val := unmarshalValue{fieldNumber: fd.GetNumber(), bytes: raw}
+		return val, nil
+
+	case proto.WireStartGroup:
+		return zeroValue, errGroupsAreNotSupported
+	default:
+		return zeroValue, proto.ErrInternalBadWireType
+	}
+}
+
+func unmarshalSimpleField(fd *desc.FieldDescriptor, v uint64) (unmarshalValue, error) {
+	val := unmarshalValue{fieldNumber: fd.GetNumber(), v: v}
+	switch fd.GetType() {
+	case dpb.FieldDescriptorProto_TYPE_BOOL,
+		dpb.FieldDescriptorProto_TYPE_UINT64,
+		dpb.FieldDescriptorProto_TYPE_FIXED64,
+		dpb.FieldDescriptorProto_TYPE_INT64,
+		dpb.FieldDescriptorProto_TYPE_SFIXED64,
+		dpb.FieldDescriptorProto_TYPE_DOUBLE:
+		return val, nil
+
+	case dpb.FieldDescriptorProto_TYPE_UINT32,
+		dpb.FieldDescriptorProto_TYPE_FIXED32:
+		if v > math.MaxUint32 {
+			return zeroValue, dynamic.NumericOverflowError
+		}
+		return val, nil
+
+	case dpb.FieldDescriptorProto_TYPE_INT32,
+		dpb.FieldDescriptorProto_TYPE_ENUM:
+		s := int64(v)
+		if s > math.MaxInt32 || s < math.MinInt32 {
+			return zeroValue, dynamic.NumericOverflowError
+		}
+		return val, nil
+
+	case dpb.FieldDescriptorProto_TYPE_SFIXED32:
+		if v > math.MaxUint32 {
+			return zeroValue, dynamic.NumericOverflowError
+		}
+		return val, nil
+
+	case dpb.FieldDescriptorProto_TYPE_SINT32:
+		if v > math.MaxUint32 {
+			return zeroValue, dynamic.NumericOverflowError
+		}
+		val.v = uint64(decodeZigZag32(v))
+		return val, nil
+
+	case dpb.FieldDescriptorProto_TYPE_SINT64:
+		val.v = uint64(decodeZigZag64(v))
```

Comment:
do we need overflow checking here?

Reply:
no because really what we're doing here is reusing the `uint64` field to store different bit patterns and then we'll re-materialize it into an int64 later (the conversion from int64 to uint64 even if it causes an overflow won't change the bit pattern so you'll still get the same results)